packages/angular_devkit/build_angular/src/angular-cli-files/utilities/find-specs.ts

```diff
@@ -0,0 +1,31 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { Path, getSystemPath, join } from '@angular-devkit/core';
+import * as glob from 'glob';
+
+export function findSpecs(input: string, workspaceRoot: Path, sourceRoot: string): string[] {
+
+  input = input.replace(/\\/g, '/'); // normalize input
+
+  // remove source root to support relative paths from root
+  // such paths are easy to get when running script via IDEs
+  if (sourceRoot && input.startsWith(sourceRoot + '/')) {
+    input = input.substr(sourceRoot.length + 1); // +1 to include slash
+  }
+
+  if (input.endsWith('.ts') && !input.endsWith('.spec.ts')) {
+    input = input.substr(0, input.length - 2) + 'spec.ts';
+  } else if (!input.endsWith('.spec.ts') && input.indexOf('*') === -1) {
+    input += '**/*.spec.ts';
```

Comment:
What is this case meant to address? Is it when the path is for a directory? We can test for that by checking the path exists and is a directly instead. I feel that's a bit safer because it assumes less. For instance, we also support `.tsx` and `.js`.

Reply:
Well, that only solves part of the problem - I would still need to know what files to look for.

Thinking more about it, it may be better to leave this for custom command in editor. 

Currently colleagues with webstorm are benefiting from this - they are able to right click a directory in files tree and run all unit test inside. It would be trivial to add `**/*.my-spec-files.ext` there.

Reply:
Can you tell me a bit more about the right click case? I don't really use webstorm myself. Is it just right click, or something like right click, copy path?

Maybe we could go with the assumption that the dir case only works for `.spec.ts` as a convenience. It doesn't seem to cause any bad cases.

Reply:
I am not a user either, but from I've seen it's:
- right click
- find external tools option
- run predefined / imported 'tool'

It basically runs a specified command and can substitute certain variables, e.g. relative path from root of selected file / directory. This is much like tasks.json for vscode, but imported from xml or manually configured somewhere in settings.

Reply:
Ok, then my request for these conditionals is to change it to this:
```
import { isDirectory } from './is-directory';
import { extname, basename } from 'path';

// ...

const matchingSpec = `${basename(input, extname(input))}.spec.ts`;

if (existsSync(matchingSpec)) {
  // Assume the user actually wants to run the spec, if there is one.
  input = matchingSpec;
} else if (isDirectory(input)) {
  // If we know it is a directory, instead assume the user wants the default spec glob inside.
  input += '**/*.spec.ts';
}
```

This way we don't exclude `.tsx` files and we verify the directory instead of something unexpected. Does this look good to you?

Reply:
It ended up a bit more complicated, but shares the same idea. Hopefully all will be good now